From 18f2a2fa599791c90943be592418b43c58f3d21c Mon Sep 17 00:00:00 2001
From: Eliad Peller <eliad@wizery.com>
Date: Sun, 8 Jun 2014 03:37:44 +0300
Subject: [PATCH 10/47] acs: skip HT_SCAN on channel sync [ACS]

Since the channel was already chosen, we don't need
to do another HT_SCAN.
This reduces some problems that occur when radar
is detected during HT_SCAN, etc.

Signed-off-by: Eliad Peller <eliad@wizery.com>
---
 src/ap/hostapd.c | 30 ++++++++++++++++++++++++++----
 1 file changed, 26 insertions(+), 4 deletions(-)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 05a61e9..f451d45 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -64,6 +64,8 @@ static int setup_interface2(struct hostapd_iface *iface);
 static void channel_list_update_timeout(void *eloop_ctx, void *timeout_ctx);
 static void hostapd_interface_setup_failure_handler(void *eloop_ctx,
 						    void *timeout_ctx);
+static int hostapd_sync_channel(struct hostapd_iface *hapd_iface);
+
 
 
 int hostapd_for_each_interface(struct hapd_interfaces *interfaces,
@@ -1592,7 +1594,18 @@ static int setup_interface2(struct hostapd_iface *iface)
 		/* Not all drivers support this yet, so continue without hw
 		 * feature data. */
 	} else {
-		int ret = hostapd_select_hw_mode(iface);
+		int ret;
+
+		if (iface->conf->ap_channel_sync) {
+			ret = hostapd_sync_channel(iface);
+			if (ret < 0)
+				return ret;
+			/* skip other checks if channel was synced */
+			if (ret)
+				goto out;
+		}
+
+		ret = hostapd_select_hw_mode(iface);
 		if (ret < 0) {
 			wpa_printf(MSG_ERROR, "Could not select hw_mode and "
 				   "channel. (%d)", ret);
@@ -1616,6 +1629,7 @@ static int setup_interface2(struct hostapd_iface *iface)
 		if (iface->conf->ieee80211h)
 			wpa_printf(MSG_DEBUG, "DFS support is enabled");
 	}
+out:
 	return hostapd_setup_interface_complete(iface, 0);
 
 fail:
@@ -1645,6 +1659,17 @@ static int hostapd_sync_channel(struct hostapd_iface *hapd_iface)
 		hapd_iface->conf->channel = iface->conf->channel;
 		hapd_iface->conf->secondary_channel =
 				iface->conf->secondary_channel;
+		/* set current mode */
+		for (i = 0; i < hapd_iface->num_hw_features; i++) {
+			struct hostapd_hw_modes *mode =
+				&hapd_iface->hw_features[i];
+
+			if (mode->mode == hapd_iface->conf->hw_mode) {
+				hapd_iface->current_mode = mode;
+				break;
+			}
+		}
+
 		wpa_printf(MSG_DEBUG, "Channel automatically synced to "
 			   "existing AP: %d (secondary: %d) (mode: %s)",
 			   iface->conf->channel,
@@ -1905,9 +1930,6 @@ static int hostapd_setup_interface_complete_sync(struct hostapd_iface *iface,
 
 	wpa_printf(MSG_DEBUG, "Completing interface initialization");
 
-	if (iface->conf->ap_channel_sync)
-		hostapd_sync_channel(iface);
-
 	if (iface->conf->channel) {
 #ifdef NEED_AP_MLME
 		int res;
-- 
1.9.1

